### ISSUE TITLE
Screen class not working on ev3dev stretch

### ISSUE THREAD

**Issue by ddemidov** [ev3dev/ev3dev-lang-python#423]
I am getting the following error with the recently released stretch package:
```
>>> import ev3dev.ev3 as ev3
>>> s = ev3.Screen()
>>> s.clear()
>>> s.update()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/ev3dev/core.py", line 3320, in update
    self.mmap[:] = self._img.tobytes("raw", "1;IR")
IndexError: mmap slice assignment is wrong size
```
It seems the same error prevents `openrobertalab` service from working:
```
$ journalctl -n 100
Oct 28 10:08:34 ev3dev openrobertalab[1992]: INFO:roberta:--- starting ---
Oct 28 10:08:34 ev3dev openrobertalab[1992]: INFO:roberta.lab:python path: /usr/bin:/usr/lib/python35.zip:/usr/lib/python3.5:/usr/lib/python3.5/plat-arm-linux-gnueabi:/usr/lib/python3.5/lib-
Oct 28 10:08:34 ev3dev openrobertalab[1992]: INFO:roberta.lab:status changed: disconnected
Oct 28 10:08:38 ev3dev openrobertalab[1992]: Traceback (most recent call last):
Oct 28 10:08:38 ev3dev openrobertalab[1992]:   File "/usr/bin/openrobertalab", line 41, in <module>
Oct 28 10:08:38 ev3dev openrobertalab[1992]:     main()
Oct 28 10:08:38 ev3dev openrobertalab[1992]:   File "/usr/bin/openrobertalab", line 36, in main
Oct 28 10:08:38 ev3dev openrobertalab[1992]:     service = Service('/org/openroberta/Lab1')
Oct 28 10:08:38 ev3dev openrobertalab[1992]:   File "/usr/lib/python3/dist-packages/roberta/lab.py", line 88, in __init__
Oct 28 10:08:38 ev3dev openrobertalab[1992]:     self.hal.clearDisplay()
Oct 28 10:08:38 ev3dev openrobertalab[1992]:   File "/usr/lib/python3/dist-packages/roberta/ev3.py", line 179, in clearDisplay
Oct 28 10:08:38 ev3dev openrobertalab[1992]:     self.lcd.update()
Oct 28 10:08:38 ev3dev openrobertalab[1992]:   File "/usr/lib/python3/dist-packages/ev3dev/core.py", line 3320, in update
Oct 28 10:08:38 ev3dev openrobertalab[1992]:     self.mmap[:] = self._img.tobytes("raw", "1;IR")
Oct 28 10:08:38 ev3dev openrobertalab[1992]: IndexError: mmap slice assignment is wrong size
Oct 28 10:08:38 ev3dev openrobertalab[1992]: INFO:roberta:--- done ---
```

This is on the latest snapshot ([snapshot-ev3dev-stretch-ev3-generic-2017-10-24.img.xz](https://oss.jfrog.org/list/oss-snapshot-local/org/ev3dev/brickstrap/2017-10-24/snapshot-ev3dev-stretch-ev3-generic-2017-10-24.img.xz)). The stretch-compatible ev3dev version is there, since leds are working:
```
>>> import ev3dev.ev3 as ev3
>>> ev3.Leds.all_off()
```

**Comment by ddemidov** [comment]
@dlech, were there any changes in framebuffer on stretch? For reference, here is what we currently do:
https://github.com/ev3dev/ev3dev-lang-python/blob/02fa6023b6b80cc6105eb951b3dd4c39757756c7/ev3dev/core.py#L3093-L3234

**Comment by dlech** [comment]
There is now 2bpp support in the driver, so `fix_info.smem_len` is probably twice as big as it was previously.

Could you change

    self.mmap[:] = self._img.tobytes("raw", "1;IR")

to

    b = self._img.tobytes("raw", "1;IR")
    self.mmap[:len(b)] = b

?

**Comment by ddemidov** [comment]
That works, although I had to invert the image (or rather stop inverting it, by using "1;R" instead of "1;IR").
The following patch works for me:
```diff
diff --git a/ev3dev/core.py b/ev3dev/core.py
index 0dd29ca..72d0f41 100644
--- a/ev3dev/core.py
+++ b/ev3dev/core.py
@@ -3317,7 +3317,8 @@ class Screen(FbMem):
         Nothing will be drawn on the screen until this function is called.
         """
         if self.var_info.bits_per_pixel == 1:
-            self.mmap[:] = self._img.tobytes("raw", "1;IR")
+            b = self._img.tobytes("raw", "1;R")
+            self.mmap[:len(b)] = b
         elif self.var_info.bits_per_pixel == 16:
             self.mmap[:] = self._img_to_rgb565_bytes()
         else:
```

I am bit surprised by the result though. `tobytes("raw", "1;R")` should give us raw 1-bit-per-pixel image. Why assigning the result to a 2-bit-per-pixel framebuffer works?

**Comment by dlech** [comment]
The fbdev device has two modes of operation, 1bpp and 2bpp. Be default, it still operates in 1bpp mode. The memory allocated has to be big enough for both modes, so the 1bpp mode only uses 1/2 of the allocated memory.

**Comment by ddemidov** [comment]
Ok, that makes sense. I'll create PRs for `stretch-stable` and `develop` then.

**Comment by dlech** [comment]
I'm wondering about the inversion though. I don't recall switching from `FB_VISUAL_MONO10` to `FB_VISUAL_MONO01` but maybe I did.

**Comment by ddemidov** [comment]
Do you need more info from me on this? The EV3 is next to me right now.

**Comment by dlech** [comment]
The driver still says `FB_VISUAL_MONO10`, so I would expect that the inversion is still needed.

https://github.com/ev3dev/lego-linux-drivers/blob/ev3dev-stretch/linux/fbdev/st7586fb.c#L483

**Comment by ddemidov** [comment]
Well, without the inversion `s.clear()` gives me black screen, and `s.draw.rectangle((10,10,168,118))` gives me a white-on-black rectangle.

**Comment by dlech** [comment]
https://github.com/ev3dev/ev3dev-lang-python/blob/8e57eeae1bc4193fba1de627b40eee2205366fc7/ev3dev/display.py#L257

says clear should fill with white, not black

**Comment by ddemidov** [comment]
I believe that is because we had to do inversion.

**Comment by ddemidov** [comment]
Fixed by #424 (`develop`, aka v2) and #425 (`stretch-stable`, aka v1). Also pushed the updated packages for the stretch-stable version.
